Inconsistency between accessing field of braced struct vs tuple struct in proc macro

If a procedural macro generates an access to a named struct field like `self.x` then the `.` may be either def_site or call_site and it works either way. But if accessing an unnamed tuple struct field like `self.0` then it only works if `.` is call_site.

I believe it should work either way in either case.

```rust
#![feature(proc_macro)]

extern crate proc_macro;
use proc_macro::{TokenStream, TokenTree, TokenNode, Spacing, Span};

#[proc_macro]
pub fn field(input: TokenStream) -> TokenStream {
    let mut iter = input.into_iter();
    let struct_name = iter.next().unwrap();
    let field_name = iter.next().unwrap();
    assert!(iter.next().is_none());

    let dot = TokenTree {
        // Call_site works for both braced structs and tuple structs.
        // Def_site works for braced structs but not tuple structs.
        span: Span::def_site(),
        kind: TokenNode::Op('.', Spacing::Alone),
    };
    vec![
        struct_name,
        dot,
        field_name,
    ].into_iter().collect()
}
```

```rust
#![feature(proc_macro)]

extern crate mac;

struct S {
    a: u8,
}

struct T(u8);

fn main() {
    let s = S { a: 0 };
    println!("{:?}", mac::field!(s a));

    let t = T(0);
    println!("{:?}", mac::field!(t 0));
}
```

```
error[E0611]: field `0` of tuple-struct `T` is private
  --> src/main.rs:55:34
   |
55 |     println!("{:?}", mac::field!(t 0));
   |                                  ^^^
```

@jseyfried

--- a/src/libsyntax/parse/parser.rs
+++ b/src/libsyntax/parse/parser.rs
@@ -2630,8 +2630,7 @@ impl<'a> Parser<'a> {
                     // A tuple index may not have a suffix
                     self.expect_no_suffix(sp, "tuple index", suf);
 
-                    let dot_span = self.prev_span;
-                    hi = self.span;
+                    let idx_span = self.span;
                     self.bump();
 
                     let invalid_msg = "invalid tuple or struct index";
@@ -2646,9 +2645,8 @@ impl<'a> Parser<'a> {
                                                     n.to_string());
                                 err.emit();
                             }
-                            let id = respan(dot_span.to(hi), n);
-                            let field = self.mk_tup_field(e, id);
-                            e = self.mk_expr(lo.to(hi), field, ThinVec::new());
+                            let field = self.mk_tup_field(e, respan(idx_span, n));
+                            e = self.mk_expr(lo.to(idx_span), field, ThinVec::new());
                         }
                         None => {
                             let prev_span = self.prev_span;

--- a/src/test/run-pass/hygiene/issue-47312.rs
+++ b/src/test/run-pass/hygiene/issue-47312.rs
@@ -1,0 +1,30 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// ignore-pretty pretty-printing is unhygienic
+
+#![feature(decl_macro)]
+#![allow(unused)]
+
+mod foo {
+    pub macro m($s:tt, $i:tt) {
+        $s.$i
+    }
+}
+
+mod bar {
+    struct S(i32);
+    fn f() {
+        let s = S(0);
+        ::foo::m!(s, 0);
+    }
+}
+
+fn main() {}

